apiserver/common/crossmodel/auth.go

```diff
@@ -67,23 +68,29 @@ type AuthContext struct {
 
 	clock              clock.Clock
 	offerThirdPartyKey *bakery.KeyPair
-	offerBakery        authentication.ExpirableStorageBakery
 
-	offerAccessEndpoint string
+	localOfferBakery authentication.ExpirableStorageBakery
+	jaasOfferBakery  authentication.ExpirableStorageBakery
+
+	localOfferAccessEndpoint string
+	jaasOfferAccessEndpoint  string
```

Comment:
We don't need this distinction. We can stay with the existing `offerBakery` and `offerAccessEndpoint`.
When the AuthContext is constructed, we pass in whichever one is needed.

We then do not need the `getBakey()` method.

Now consider the places where we do this
`if a.jaasOfferBakery != nil {`

1. create the discharge macaroon
we can just have methods on each bakery to do this - embed the ExpirableStorageBakery inside an interface which has the create macaroon method

2. setting the offer uuid to ""
on the above (new) interface, we can just have a `getDeclaredCaveats()` method (that's the only difference), maybe using a helper for the common bits

3. inferDeclaredFromMacaroon
again, add a method to the interface

So we might have

```
type offerBakery interface {
  authentication.ExpirableStorageBakery
  createDischargeMacaroon(ctx context.Context, username string, requiredValues, declaredValues map[string]string, op bakery.Op, version bakery.Version) (*bakery.Macaroon, error)
  getDelcaredCaveats(username, offerUUID string, sourceModelTag names.ModelTag, offerExpiryTime time.Time) []checkers.Caveat
  inferDeclaredFromMacaroon(mac macaroon.Slice, requiredValues map[string]string) map[string]string
}
```

Reply:
Maybe there's an even better way, maybe we can get it down to one or two methods on some interface like the above....